### DISCUSSION TITLE
Make a csv file from a class that derives from “Material”

### DISCUSSION THREAD

**Discussion by styyokuda** [idaholab/moose#24222] [Q&A General] [answered]
I am writing a class that derives from “Material.”
My program includes such as:
```
.
_length(declareProperty<Real>("length")),
_temperature(coupledValue("heat_variable"))
.

float some_constant = 100;
_length[_qp] = some_constant * _temperature[_qp];
```
Now, I would like to generate a csv file includes member variable “_length[_qp].”

Please help.
S. Thomas

**Comment by GiudGiud** [comment, marked as answer]
hello

Just set the output_properties and outputs parameter in the input for this Material.
Or do you want "output" to always be done?
```
[Materials]
[ my _mat ]
   ...
   output_properties= 'length'
    outputs = 'csv'
  []
[Outputs]
  [csv]
    type = CSV
  []
[]
```

**Reply by styyokuda** [reply]
Thank you GiudGiudty,
Are you suggesting to use "type = OutputTestMaterial?"
Yes, I would like it done always.
Sincerely,
S. Thomas

**Reply by GiudGiud** [reply]
no the type is the type of your Material. 

Always might be difficult actually, because the `csv` output block might not exist.
You could set the default of outputs to 'all' ?

**Reply by styyokuda** [reply]
For a quick check, I am using the below.
It did not make a csv file but the 'diff_u' is included in the exodus file.
A csv file is supposed to be generated, isn't it?
Please help.
```
[Materials]
  [./conduct_u]
    type = GenericConstantMaterial
    prop_names = 'diff_u'
    prop_values = 1.0
    block = 0
    output_properties = 'diff_u'
    outputs = 'csv'
  [../]
[Outputs]
  exodus = true
  [csv]
    type = CSV
  []
[]
```

**Reply by styyokuda** [reply]
I remember that the Elemental variable family is not shown in a csv file. So, 'diff_u' may have the Elemental variable family. I will work on my "_length[_qp]" to see if I can print it in a csv file.
Sincerely, 
S. Thomas

**Reply by GiudGiud** [reply]
oh  somehow I did not connect that this CSV output wasnt going to get set up automatically. 
This wont work. Let me find a different solution

**Reply by GiudGiud** [reply]
This is the object you want to use
https://mooseframework.inl.gov/source/vectorpostprocessors/MaterialVectorPostprocessor.html

**Reply by styyokuda** [reply]
Thank you, Guillaume!
Actually I would like to make a CSV file for values along boundary lines. In that case, do you recommend to use "LineValueSampler" or any other methods like giving the boundary numbers (or names) instead of coordinates.  Please help. S. Thomas

**Reply by GiudGiud** [reply]
If the boundaries are straight lines, then LineValueSampler is the one to go with.

If not, then we do not have a BoundaryValueSampler. 
However, if you are OK with the odd workflow you could:
- use LowerDBlockFromSidesetGenerator on the boundary of interest
- use the ElementValueSampler and block restrict it to that lower D block
